src/stim/simulators/frame_simulator.pybind.cc

```diff
@@ -770,4 +770,91 @@ void stim_pybind::pybind_frame_simulator_methods(
                 [stim.PauliString("+YX__")]
         )DOC")
             .data());
+
+    c.def(
+        "apply_pauli_errors",
+        [](FrameSimulator<MAX_BITWORD_WIDTH> &self,
+           const pybind11::object &pauli,
+           const pybind11::object &mask
+           ) {
+            uint8_t p = 255;
+            try {
+                p = pybind11::cast<uint8_t>(pauli);
+            } catch (const pybind11::cast_error &) {
+                try {
+                    std::string s = pybind11::cast<std::string>(pauli);
+                    if (s == "X") {
+                        p = 1;
+                    } else if (s == "Y") {
+                        p = 2;
+                    } else if (s == "Z") {
+                        p = 3;
+                    } else if (s == "I" || s == "_") {
+                        p = 0;
+                    }
+                } catch (const pybind11::cast_error &) {
+                }
+            }
+
+            bool flip_z_part = p & 2;
+            bool flip_x_part = 6 >> p & 1;  // 0b0110 >> p & 0b0001
+
+            if (pybind11::isinstance<pybind11::array_t<bool>>(mask)) {
+                // can use pybind11::isinstance<pybind11::array_t<bool, pybind11::array::c_style>>
+                // to check for dense c_ordered array for copying, if this is too slow for you
+                const pybind11::array_t<bool> &arr = pybind11::cast<pybind11::array_t<bool>>(mask);
+                if (arr.ndim() == 2) {
+                    //pybind11::ssize_t???
+                    size_t major = arr.shape(0);
+                    size_t minor = arr.shape(1);
+                    //size_t major_stride = arr.strides(0);
+                    //size_t minor_stride = arr.strides(1);
+                    auto u = arr.unchecked<2>();
+                    for (size_t i = 0; i < major; i++){
+                        for (size_t j = 0; j < minor; j++){
+                            auto b = u.data(i, j);
+                            self.x_table[i][j] ^= *b & flip_x_part;
+                            self.z_table[i][j] ^= *b & flip_z_part;
+                        }
+                    }
+                } else {
+                    throw std::invalid_argument(
+                    "apply_pauli_errors currently only supports a mask that is a 2D bool numpy array");
+                }
+            } else {
+                    throw std::invalid_argument(
```

Comment:
Try to merge the two throws since they're the same error message.

Reply:
Not sure how to do this, given i need to do the dim=2 check after i've done the cast, which means ive already done the is_array check; I could change the reference to a pointer so i can have it be NULL before I do the cast? (not sure if this is frowned upon in c++ land) or i could do the cast check twice, or i could allocate extra variables,
Seems like more of a style thing here, so happy to be told what to do

Reply:
Not a big deal if it's hard to do.